File: ios/Classes/Data/AppInboxCoder.swift
//
//  AppInboxCoder.swift
//  exponea
//
//  Created by Adam Mihalik on 03/03/2023.
//

import Foundation
import ExponeaSDK

class AppInboxCoder {
    static func encode(_ source: MessageItem) throws -> [String:Any?] {
        // keep in sync with app_inbox.dart
        return [
            "id": source.id,
            "type": source.type,
            "isRead": source.read,
            "createTime": source.rawReceivedTime,
            "content": try self.normalizeData(source.rawContent)
        ]
    }
    
    private static func normalizeData(_ source: [String: JSONValue]?) throws -> [String: Any?] {
        let rawContent = try JSONEncoder().encode(source)
        let normalized = try JSONSerialization.jsonObject(
            with: rawContent, options: []
        ) as? [String: Any?]
        return normalized ?? [:]
    }
}

File: android/src/main/kotlin/com/exponea/data/AppInboxCoder.kt
package com.exponea.data

import com.exponea.getNullSafely
import com.exponea.sdk.models.MessageItem
import com.exponea.sdk.models.MessageItemAction
import com.exponea.sdk.models.MessageItemAction.Type

class AppInboxCoder {
    companion object {
        fun encode(source: MessageItem): Map<String, Any?> {
            return mapOf(
                "id" to source.id,
                "type" to source.rawType,
                "isRead" to source.read,
                "createTime" to source.receivedTime,
                "content" to source.rawContent
            )
        }

        fun decodeAction(source: Map<String, Any?>): MessageItemAction? {
            val sourceType = Type.find(source.getNullSafely("action")) ?: return null
            return MessageItemAction().apply {
                type = sourceType
                title = source.getNullSafely("title")
                url = source.getNullSafely("url")
            }
        }
    }
}